File: 0431_encode_n-ary_tree/solution.java
class Node {
    public int val;
    public List<Node> children;

    public Node() {
    }

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};

public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}

class Codec {

    // Encodes an n-ary tree to a binary tree.
    public TreeNode encode(Node root) {
        if (root == null) {
            return null;
        }
        TreeNode res = new TreeNode(root.val);
        if (root.children.size() > 0) {
            res.left = encode(root.children.get(0));
        }
        TreeNode curr = res.left;
        for (int i = 1; i < root.children.size(); i++) {
            curr.right = encode(root.children.get(i));
            curr = curr.right;
        }
        return res;
    }

    // Decodes your binary tree to an n-ary tree.
    public Node decode(TreeNode root) {
        if (root == null) {
            return null;
        }
        Node res = new Node(root.val, new LinkedList<>());
        TreeNode curr = root.left;
        while (curr != null) {
            res.children.add(decode(curr));
            curr = curr.right;
        }
        return res;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.decode(codec.encode(root));

File: 0431_encode_n-ary_tree/solution.py
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Codec:

    # Encodes an n-ary tree to a binary tree.
    def encode(self, root: 'Node') -> TreeNode:
        if not root:
            return

        res = TreeNode(root.val)
        if not root.children:
            return res

        res.left = self.encode(root.children[0])
        curr = res.left
        for child in root.children[1:]:
            curr.right = self.encode(child)
            curr = curr.right
        return res

    # Decodes your binary tree to an n-ary tree.
    def decode(self, data: TreeNode) -> 'Node':
        if not data:
            return

        res = Node(data.val, [])
        node = data.left
        while node:
            res.children.append(self.decode(node))
            node = node.right
        return res

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.decode(codec.encode(root))
